L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc

```diff
@@ -0,0 +1,286 @@
+#include "DataFormats/L1TCorrelator/interface/TkPrimaryVertex.h"
+#include "DataFormats/L1TCorrelator/interface/TkHTMiss.h"
+#include "DataFormats/L1TCorrelator/interface/TkHTMissFwd.h"
+#include "DataFormats/L1Trigger/interface/EtSum.h"
+#include "DataFormats/L1Trigger/interface/TkJetWord.h"
+#include "L1Trigger/L1TTrackMatch/interface/L1TkHTMissEmulatorProducer.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+using namespace l1t;
+
+class L1TkHTMissEmulatorProducer : public edm::stream::EDProducer<> {
+public:
+  explicit L1TkHTMissEmulatorProducer(const edm::ParameterSet&);
+  ~L1TkHTMissEmulatorProducer() override;
+
+private:
+  virtual void beginJob();
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  virtual void endJob(); 
+
+  // ----------member data ---------------------------
+
+  bool debug_ = false;
+  bool saveLUTs_ = false;
+  bool displaced_;
+
+  // floats used for debugging
+  float jetMinPt__;            // [GeV]
+  float jetMaxEta__;           // [rad]
+  int minNtracksHighPt__;
+  int minNtracksLowPt__;
+
+  l1tmhtemu::pt_t jetMinPt_;
+  l1tmhtemu::eta_t jetMaxEta_;
+  l1tmhtemu::ntracks_t minNtracksHighPt_;
+  l1tmhtemu::ntracks_t minNtracksLowPt_;
+
+  std::vector<l1tmhtemu::phi_t> cosLUT_;
+  std::vector<l1tmhtemu::MHTphi_t> atanLUT_;
+  std::vector<l1tmhtemu::Et_t> magNormalisationLUT_;
+
+  std::string L1MHTCollectionName_;
+
+  const edm::EDGetTokenT<TkJetWordCollection> jetToken_;
+};
+
+L1TkHTMissEmulatorProducer::L1TkHTMissEmulatorProducer(const edm::ParameterSet& iConfig)
+  : jetToken_(consumes<TkJetWordCollection>(iConfig.getParameter<edm::InputTag>("L1TkJetEmulationInputTag"))) {
+  debug_ = (bool)iConfig.getParameter<bool>("debug");
+  saveLUTs_ = (bool)iConfig.getParameter<bool>("saveLUTs");
+  displaced_ = iConfig.getParameter<bool>("displaced");
+
+  // floats used for debugging
+  jetMinPt__ = (float)iConfig.getParameter<double>("jet_minPt");
+  jetMaxEta__ = (float)iConfig.getParameter<double>("jet_maxEta");
+  minNtracksHighPt__ = iConfig.getParameter<int>("jet_minNtracksHighPt");
+  minNtracksLowPt__ = iConfig.getParameter<int>("jet_minNtracksLowPt");
+
+  jetMinPt_ = l1tmhtemu::digitizeSignedValue<l1tmhtemu::pt_t>((float)iConfig.getParameter<double>("jet_minPt"), l1tmhtemu::kInternalPtWidth, l1tmhtemu::kStepPt);
+  jetMaxEta_ = l1tmhtemu::digitizeSignedValue<l1tmhtemu::eta_t>((float)iConfig.getParameter<double>("jet_maxEta"), l1tmhtemu::kInternalEtaWidth, l1tmhtemu::kStepEta);
+  minNtracksHighPt_ = (l1tmhtemu::ntracks_t)iConfig.getParameter<int>("jet_minNtracksHighPt");
+  minNtracksLowPt_ = (l1tmhtemu::ntracks_t)iConfig.getParameter<int>("jet_minNtracksLowPt");
+
+  int cosLUTbins = floor(l1tmhtemu::kMaxCosLUTPhi / l1tmhtemu::kStepPhi);
+  cosLUT_ = l1tmhtemu::generateCosLUT(cosLUTbins);
+
+  // save LUTs
+  if(saveLUTs_){
+
+    l1tmhtemu::writeLUTtoFile<l1tmhtemu::phi_t>(cosLUT_, "cos", ",");
+
+    int aSteps = 8;
```

Comment:
Why is this defined in both the `produce` method and the constructor? It's a constant, so it should be defined once in the class declaration.

Reply:
fixed in the latest commit.